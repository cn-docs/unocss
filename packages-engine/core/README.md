# @unocss/core

The core engine of [UnoCSS](https://github.com/unocss/unocss) without any presets. It can be used as the engine of your own atomic CSS framework.

## Documentation

Please refer to the [documentation](https://unocss.dev/tools/core).

## License

MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
