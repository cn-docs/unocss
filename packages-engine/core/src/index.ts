export * from './config'
export * from './constants'
export * from './extractors'
export * from './generator'
export * from './types'
export * from './utils'
