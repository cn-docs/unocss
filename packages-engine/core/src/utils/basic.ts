export function toArray<T>(value: T | T[] = []): T[] {
  return Array.isArray(value) ? value : [value]
}

export function uniq<T>(value: T[]): T[] {
  return Array.from(new Set(value))
}

export function uniqueBy<T>(array: readonly T[], equalFn: (a: T, b: T) => boolean): T[] {
  return array.reduce((acc: T[], cur: T) => {
    const index = acc.findIndex((item: T) => equalFn(cur, item))
    if (index === -1)
      acc.push(cur)
    return acc
  }, [])
}

export function isString(s: any): s is string {
  return typeof s === 'string'
}
