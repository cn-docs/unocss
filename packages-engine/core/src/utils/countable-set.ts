export class CountableSet<K> extends Set<K> {
  _map: Map<K, number>

  constructor(values?: Iterable<K>) {
    super(values)
    this._map ??= new Map()
  }

  add(key: K) {
    this._map ??= new Map()
    this._map.set(key, (this._map.get(key) ?? 0) + 1)
    return super.add(key)
  }

  delete(key: K) {
    this._map.delete(key)
    return super.delete(key)
  }

  clear() {
    this._map.clear()
    super.clear()
  }

  getCount(key: K) {
    return this._map.get(key) ?? 0
  }

  setCount(key: K, count: number) {
    this._map.set(key, count)
    return super.add(key)
  }
}

export function isCountableSet<T = string>(value: any): value is CountableSet<T> {
  return value instanceof CountableSet
}
