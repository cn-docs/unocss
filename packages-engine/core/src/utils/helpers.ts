import type { ParsedUtil, RawUtil, StringifiedUtil, Variant, VariantObject } from '../types'

export const attributifyRE = /^\[(.+?)~?="(.*)"\]$/
export const cssIdRE = /\.(css|postcss|sass|scss|less|stylus|styl)($|\?)/
// eslint-disable-next-line regexp/no-obscure-range
export const validateFilterRE = /[\w\u00A0-\uFFFF%-?]/

export function isAttributifySelector(selector: string) {
  return selector.match(attributifyRE)
}

export function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}

export function normalizeVariant(variant: Variant<any>): VariantObject<any> {
  return typeof variant === 'function'
    ? { match: variant }
    : variant
}

export function isRawUtil(util: ParsedUtil | RawUtil | StringifiedUtil): util is RawUtil {
  return util.length === 3
}

export function notNull<T>(value: T | null | undefined): value is T {
  return value != null
}

export function noop() {}
