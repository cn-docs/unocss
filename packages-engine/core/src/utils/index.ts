export * from './basic'
export * from './countable-set'
export * from './escape'
export * from './events'
export * from './helpers'
export * from './layer'
export * from './map'
export * from './object'
export * from './variant-group'
export * from './warn'
