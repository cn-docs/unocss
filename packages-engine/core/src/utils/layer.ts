import type { Rule } from '../types'

export function withLayer<T extends object>(layer: string, rules: Rule<T>[]) {
  rules.forEach((r) => {
    if (!r[2])
      r[2] = { layer }
    else
      r[2].layer = layer
  })
  return rules
}
