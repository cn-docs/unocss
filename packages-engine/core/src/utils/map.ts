export class TwoKeyMap<K1, K2, V> {
  _map = new Map<K1, Map<K2, V>>()

  get(key1: K1, key2: K2): V | undefined {
    const m2 = this._map.get(key1)
    if (m2)
      return m2.get(key2)
  }

  getFallback(key1: K1, key2: K2, fallback: V): V {
    let m2 = this._map.get(key1)
    if (!m2) {
      m2 = new Map<K2, V>()
      this._map.set(key1, m2)
    }
    if (!m2.has(key2))
      m2.set(key2, fallback)
    return m2.get(key2)!
  }

  set(key1: K1, key2: K2, value: V) {
    let m2 = this._map.get(key1)
    if (!m2) {
      m2 = new Map()
      this._map.set(key1, m2)
    }
    m2.set(key2, value)
    return this
  }

  has(key1: K1, key2: K2) {
    return this._map.get(key1)?.has(key2)
  }

  delete(key1: K1, key2: K2) {
    return this._map.get(key1)?.delete(key2) || false
  }

  deleteTop(key1: K1) {
    return this._map.delete(key1)
  }

  map<T>(fn: (v: V, k1: K1, k2: K2) => T): T[] {
    return Array.from(this._map.entries())
      .flatMap(([k1, m2]) => Array.from(m2.entries())
        .map(([k2, v]) => {
          return fn(v, k1, k2)
        }))
  }
}

export class BetterMap<K, V> extends Map<K, V> {
  getFallback(key: K, fallback: V): V {
    const v = this.get(key)
    if (v === undefined) {
      this.set(key, fallback)
      return fallback
    }
    return v
  }

  map<R>(mapFn: (value: V, key: K) => R): R[] {
    const result: R[] = []
    this.forEach((v, k) => {
      result.push(mapFn(v, k))
    })
    return result
  }

  flatMap<R extends readonly unknown[]>(mapFn: (value: V, key: K) => R): R[number][] {
    const result: R[number][] = []
    this.forEach((v, k) => {
      result.push(...mapFn(v, k))
    })
    return result
  }
}
