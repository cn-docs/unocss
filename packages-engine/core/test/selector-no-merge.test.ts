import { createGenerator } from '@unocss/core'
import { variantMatcher } from '@unocss/preset-mini/utils'
import { describe, expect, it } from 'vitest'

describe('selector', async () => {
  const uno = await createGenerator({
    rules: [
      [/^to-merge$/, () => [{ merged: 1 }]],
      [/^merge-candidate$/, () => ({ merged: 1 })],
      [/^not-merged$/, () => ({ merged: 1 }), { noMerge: true }],
    ],
  })

  it('rules split selector', async () => {
    const { css } = await uno.generate('to-merge merge-candidate not-merged')
    expect(css).toMatchSnapshot()
  })
})

describe('variant', async () => {
  const uno = await createGenerator({
    shortcuts: [
      [/^m1-(.+)$/, ([, s]) => `moz:${s} webkit:${s}`],
      [/^m2-(.+)$/, ([, s]) => `moz:${s} merge-candidate`],
      [/^m3-(.+)$/, ([, s]) => `moz:${s} merge-candidate-early`],
    ],
    variants: [
      variantMatcher('moz', s => ({ pseudo: `${s.pseudo}::non-breaking` })),
      variantMatcher('webkit', s => ({ pseudo: `${s.pseudo}::breaking` })),
    ],
    rules: [
      [/^no-merge$/, () => ({ merged: 1 }), { noMerge: true }],
      ['merge-candidate', { merged: 1 }],

      ['merge-candidate-early', { late: 2 }],
      [/^no-merge-ordered$/, () => ({ late: 2 }), { noMerge: true }],
    ],
  })

  it('variant split selector', async () => {
    const { css } = await uno.generate('moz:no-merge webkit:no-merge')
    expect(css).toMatchSnapshot()
  })

  it('variant split shortcuts', async () => {
    const { css } = await uno.generate('m1-no-merge')
    expect(css).toMatchSnapshot()
  })

  it('variant shortcuts early', async () => {
    const { css } = await uno.generate('m2-no-merge')
    expect(css).toMatchSnapshot()
  })

  it('variant shortcuts late', async () => {
    const { css } = await uno.generate('m3-no-merge-ordered')
    expect(css).toMatchSnapshot()
  })
})
