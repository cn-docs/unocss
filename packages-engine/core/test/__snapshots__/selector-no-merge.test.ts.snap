// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selector > rules split selector 1`] = `
"/* layer: default */
.merge-candidate,
.to-merge{merged:1;}
.not-merged{merged:1;}"
`;

exports[`variant > variant shortcuts early 1`] = `
"/* layer: shortcuts */
.m2-no-merge::non-breaking{merged:1;}
.m2-no-merge{merged:1;}"
`;

exports[`variant > variant shortcuts late 1`] = `
"/* layer: shortcuts */
.m3-no-merge-ordered{late:2;}
.m3-no-merge-ordered::non-breaking{late:2;}"
`;

exports[`variant > variant split selector 1`] = `
"/* layer: default */
.moz\\:no-merge::non-breaking{merged:1;}
.webkit\\:no-merge::breaking{merged:1;}"
`;

exports[`variant > variant split shortcuts 1`] = `
"/* layer: shortcuts */
.m1-no-merge::breaking{merged:1;}
.m1-no-merge::non-breaking{merged:1;}"
`;
